### PR TITLE
fix #10305 nim cpp is now nan-correct at CT

### LINKED ISSUE #10305: nim cpp treats Nan as 0.0 (during compile time)
```
nim cpp -d:release -o:bin/nim_cpp compiler/nim.nim
bin/nim_cpp cpp -r main.nim
```

```
CT
nan
nan
nan
true
nan
true
nan
RT
0.0
nan
nan
true
nan
false
nan
```

```nim
import math

proc main()=
    let a = Nan
    let b = $a
    echo b # 0.0 BUG

    echo $Nan # nan
    let a2 = $Nan
    echo a2  # nan
    echo 0.0/0.0 == 0.0 # true BUG
    echo 0.0/0.0 # nan

    let a3 = 0.0
    echo a3/0.0 == 0.0 # false
    echo a3/0.0 # nan

static: echo "CT"
static: main()
echo "RT"
main()

```

this explains oddly failing CI tests in cpp mode

```
FAIL: tests/float/tfloatmod.nim C++
Test "tests/float/tfloatmod.nim" in category "float"
Failure: reExitcodesDiffer
Expected:
exitcode: 0

Gotten:
exitcode: 1

Output:
tfloatmod.nim(129)       tfloatmod
tfloatmod.nim(126)       main
tfloatmod.nim(106)       testSpecialCases
system.nim(3978)         failedAssertImpl
system.nim(3971)         raiseAssert
system.nim(3011)         sysFatal
Error: unhandled exception: /Users/travis/build/nim-lang/Nim/tests/float/tfloatmod.nim(106, 13) `(r == expected) or (r.isnan and expected.isnan)` -inf mod -inf == nan but expected 0.0 [AssertionError]
```

marking as high priority as it affects running cpp mode in CI normally (right now it's being run as allowed failure, which means if new regressions happen with cpp mode, we won't know): indeed it seems like root cause for 3 failures in cpp mode out of ~6:

* tests/float/tfloatnan.nim
* lib/pure/math.nim
* tests/float/tfloatmod.nim

## EDIT:
* looks like the bug is during const folding and  compile time

### PR DESCRIPTION
**Pull request by timotheecour**
* fix #10305 nim cpp is now nan-correct at CT #10310 and added a test that would fail before PR
* should fix these failing tests in CI (with nim cpp)
  * tests/float/tfloatnan.nim
  * lib/pure/math.nim
  * tests/float/tfloatmod.nim

I finally figured out root cause of #10305, but it was a rather painful debug session

Here's a minimal test case to explain what's going on:
```cpp
#include <stdio.h>

#define CASE_VOLATILE // otherwise const folding optimizer kicks in
int main (int argc, char *argv[]) {
  #ifdef CASE_VOLATILE
    volatile double a1 = 0.0;
  #else
    double a1 = 0.0;
  #endif
  double a2 = 0.0;
  double a3 = a1 / a2;
  int t = a3 == 0.0;
  int t2 = a3 == a3;
  int t3 = a3 != a3;
  printf("a3=%g,t=%d,t2=%d,t3=%d\n", a3, t, t2, t3);
  return 0;
}
```

```
clang++ -o /tmp/D20181107T193702 -ffast-math -O3 $timn_D/tests/nim/all/t0103.cpp
/tmp/D20181107T193702
a3=nan,t=1,t2=1,t3=0 # BUG

without  -ffast-math
a3=nan,t=0,t2=0,t3=1
```

so `-ffast-math` affects Nan correctness;
note that :

`$nimc_D/config/nim.cfg` contains
```
clang.options.speed = "-O3"
```
so that explains why this issue was only affecting `nim cpp` (since $nimc_D/compiler/extccomp.nim defined `optSpeed: " -O3 -ffast-math "` for both c and cpp, but `c` was overridden by `$nimc_D/config/nim.cfg`)

## note
there's a tradeoff between speed and accuracy at play here; but `-d:release` should be correct by default; but I added a note in doc/nimc.rst regarding more aggressive options for speed, including `--passC:-ffast-math`

### CODE REVIEW

File: tests/float/tfloatnan.nim [hunk 1]
```diff
@@ -14,3 +14,20 @@ echo "Nim: ", f32, " (float)"
...
+
+proc fun() =
+  # issue #10305
+  # with `-O3 -ffast-math`, generated C/C++ code is not nan compliant
+  # user can pass `--passC:-ffast-math` if he doesn't care.
+  let a1 = 0.0
+  let a = 0.0/a1
+  let b1 = a == 0.0
+  let b2 = a == a
+  let b3 = a != a
```
Comment (Araq) [inline tests/float/tfloatnan.nim]: ``b3`` is unused!

Comment (timotheecour) [reply tests/float/tfloatnan.nim]: fixed

**Comment by timotheecour** [comment]
PTAL, fixed `b3 unused`

**Comment by nc-x** [comment]
Won't the user still have the issue when he uses the optimization? So then, how is this considered a bugfix?

**Comment by timotheecour** [comment]
PTAL

**Pull request merged**